Stub of Dart 3 breaking change (#7931)
flutter/website: src/release/breaking-changes/dart-3.md

---
title: Dart 3 API removals & no unsound null safety support.

> description: Dart 3 will remove several deprecated APIs, and the ability to use unsound null safety.

Reviewer: ```suggestion
description: Dart 3 removes several deprecated APIs, and the ability to use unsound null safety.
```

---

## Summary

> The Dart language is preparing for a new major release, [Dart 3], in which
> there will be several breaking changes:
> 
> * **No unsound null safety**: All code must run with sound null safety.
> * **Removal of deprecated core library APIs**: Several currently deprecated API
>   members in Dart core libraries are being removed.

Reviewer: ```suggestion
The Dart language is preparing for a new major release, [Dart 3][], in which
there are several breaking changes:

* **No unsound null safety**: All code must run with _sound_ null safety.
* **Removal of deprecated core library APIs**: Several deprecated API
   members in Dart core libraries are being removed.
```

## Timeline

The Dart team expects this release by mid 2023. The Flutter team expects to
include this release in a Flutter SDK release around the same time.

## No unsound null safety

The Dart team is removing support for unsound null safety in Dart 3: 
[Dart breaking change #49530]. Concretely this implies that Dart 3 will **not
support**:

1. Code using an SDK constraint where the min-constraint isn't at least 2.12

>    (e.g. `sdk: ">=2.7.0"` will be disallowed and won't resolve).

Reviewer: ```suggestion
   (for example, `sdk: ">=2.7.0"` will be disallowed and won't resolve).
```

> 1. Libraries using [per-library language version markers] selecting versions
>    less than 2.12 (e.g. `// @dart=2.9` will be disallowed).

Reviewer: ```suggestion
1. Libraries using [per-library language version markers][] selecting versions
   less than 2.12 (for example, `// @dart=2.9` will be disallowed).
```

### Backwards compatibility for null safe packages

Dart 3 sound null safety will be **backwards compatible** with with code already
migrated to use null safety in language versions 2.12 and later.
 
This means that a Dart 3.0 SDK will allow SDK constraints where the lower-bound
is at least `2.12`, even when the SDK upper-bound only allows versions less than
3 (`<3.0`). For example, a package with the following constraint will be allowed
to resolve with a Dart 3.x SDK:

```
environment:
  sdk: '>=2.12.0 <3.0.0'
```

This allows developers to use Dart 3 sound null safety with packages that have been migrated to 2.12 null safety.

### Migration guide

> All code must be migrated to use full sound null safety. For details, see the
> [Dart null safety migration guide].

Reviewer: ```suggestion
All code must be migrated to use full sound null safety.
For details, see the [Dart null safety migration guide][].
```

## Removal of deprecated core library APIs

Several currently `@deprecated` API members in the Dart core libraries are being
removed in Dart 3.

We believe these changes to have low impact on code migrated to use null safety.

> When we release the first Dart 3 alpha build -- in early 2032 -- you will be

Reviewer 2: nit: 2023 :-)

> When we release the first Dart 3 alpha build -- in early 2032 -- you will be

Reviewer: ```suggestion
When we release the first Dart 3 alpha build -- in early 2023 -- you will be
```
